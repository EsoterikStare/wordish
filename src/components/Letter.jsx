import React from 'react';
import { string } from 'prop-types';
import { Grid, styled } from '@mui/material';

/**
 * TODO: Add little "pop" animation when adding a letter
 * TODO: Add card flip animation on submit
 */
const StyledDiv = styled('div')(({ theme, type }) => ({
  alignItems: 'center',
  backgroundColor: theme.palette.letterStatus[type],
  border: '2px solid',
  borderColor: type ? 'transparent' : '#333',
  borderRadius: 2,
  color: 'white',
  display: 'flex',
  fontSize: 32,
  fontFamily: 'sans-serif',
  height: 50,
  justifyContent: 'center',
  width: 50,
  userSelect: 'none',
}));

const Letter = ({ guessChar, type }) => (
  <Grid id="letter-grid" item>
    <StyledDiv type={type}>{guessChar.toUpperCase()}</StyledDiv>
  </Grid>
);

Letter.propTypes = {
  guessChar: string.isRequired,
  type: string,
};

Letter.defaultProps = {
  type: undefined,
};

export default Letter;
