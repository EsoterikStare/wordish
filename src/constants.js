/* eslint-disable import/prefer-default-export */

export const letterStatus = {
  LOCATED: 'located',
  PRESENT: 'present',
  ABSENT: 'absent'
};
