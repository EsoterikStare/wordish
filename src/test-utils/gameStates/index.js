import loseGameStateJSON from './loseGameState.json';
import midGuessGameStateJSON from './midGuessGameState.json';
import newGameStateJSON from './newGameState.json';
import winGameStateJSON from './winGameState.json';

export const loseGameState = JSON.stringify(loseGameStateJSON);
export const midGuessGameState = JSON.stringify(midGuessGameStateJSON);
export const newGameState = JSON.stringify(newGameStateJSON);
export const winGameState = JSON.stringify(winGameStateJSON);
