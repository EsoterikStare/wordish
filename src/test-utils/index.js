export * from './gameStates';
export { default as setup } from './setup';
