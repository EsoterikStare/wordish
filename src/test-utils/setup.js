import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { node } from 'prop-types';
import React from 'react';

import { AppStateProvider } from '../AppStateContext';
import AppThemeProvider from '../AppThemeProvider';

const Wrapper = ({ children }) => (
  <AppStateProvider>
    <AppThemeProvider>{children}</AppThemeProvider>
  </AppStateProvider>
);
Wrapper.propTypes = {
  children: node.isRequired,
};
Wrapper.defaultProps = {};

const setup = jsx => {
  const { rerender, ...rest } = render(<Wrapper>{jsx}</Wrapper>);
  return {
    user: userEvent.setup(),
    rerender: newJsx => rerender(<Wrapper>{newJsx}</Wrapper>),
    ...rest,
  };
};

export default setup;
