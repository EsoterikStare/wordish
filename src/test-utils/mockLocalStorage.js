import { newGameState } from './gameStates';

const mockLocalStorage = (gameState = newGameState) => {
  const originalLocalStorage = { ...window.localStorage };
  const mockValues = {
    clear: jest.fn(),
    // Provide a known win state from localStorage mock that will be used by AppStateProvider.
    getItem: jest.fn(() => gameState),
    removeItem: jest.fn(),
    setItem: jest.fn(),
  };

  // Mock window.localStorage with assertable functions.
  Object.defineProperty(window, 'localStorage', {
    value: mockValues,
    writable: true,
  });

  const resetLocalStorage = () => {
    Object.defineProperty(window, 'localStorage', {
      value: originalLocalStorage,
    });
  };

  return {
    resetLocalStorage,
    ...mockValues,
  };
};

export default mockLocalStorage;
