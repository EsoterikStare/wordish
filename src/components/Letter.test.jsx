import { screen } from '@testing-library/react';
import React from 'react';

import Letter from './Letter';

import { setup } from '../test-utils';
import { letterStatus } from '../constants';

describe('Letter', () => {
  it('should be defined', () => {
    expect(Letter).toBeDefined();
  });
  it('should render', () => {
    setup(<Letter guessChar="a" />);
    expect(screen.getByText(/a/i)).toBeInTheDocument();
  });
  it('should always render a capital letter', () => {
    setup(<Letter guessChar="a" />);
    expect(screen.getByText('A')).toBeInTheDocument();
  });
  it('should look correct for a guessed value', () => {
    setup(<Letter guessChar="a" />);
    expect(screen.getByTestId('letter-grid')).toMatchSnapshot();
  });
  it('should look correct for an absent value', () => {
    setup(<Letter guessChar="a" type={letterStatus.ABSENT} />);
    expect(screen.getByTestId('letter-grid')).toMatchSnapshot();
  });
  it('should look correct for an present value', () => {
    setup(<Letter guessChar="a" type={letterStatus.PRESENT} />);
    expect(screen.getByTestId('letter-grid')).toMatchSnapshot();
  });
  it('should look correct for an located value', () => {
    setup(<Letter guessChar="a" type={letterStatus.LOCATED} />);
    expect(screen.getByTestId('letter-grid')).toMatchSnapshot();
  });
});
