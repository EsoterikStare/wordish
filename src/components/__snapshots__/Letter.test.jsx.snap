// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Letter should look correct for a guessed value 1`] = `
<div
  class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
  data-testid="letter-grid"
>
  <div
    class="css-1i1qu0k"
  >
    A
  </div>
</div>
`;

exports[`Letter should look correct for an absent value 1`] = `
<div
  class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
  data-testid="letter-grid"
>
  <div
    class="css-n6nm69"
    type="absent"
  >
    A
  </div>
</div>
`;

exports[`Letter should look correct for an located value 1`] = `
<div
  class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
  data-testid="letter-grid"
>
  <div
    class="css-1f303dk"
    type="located"
  >
    A
  </div>
</div>
`;

exports[`Letter should look correct for an present value 1`] = `
<div
  class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
  data-testid="letter-grid"
>
  <div
    class="css-12zm7ch"
    type="present"
  >
    A
  </div>
</div>
`;
